### PR TITLE
8963-globalstorage-improvement-for-sqltools

### LINKED ISSUE #8963: Request 'setStorageValue' failed on closing file without workspace when SQLTools extension is install
### Bug Description:

When I install SQLTools extension and close file without workspace I get this error: (attached screen video)
```
root ERROR [hosted-plugin: 12752] Promise rejection not handled in one second: Error: Request setStorageValue failed with message: Cannot save data: no opened workspace , reason: Error: Request setStorageValue failed with message: Cannot save data: no opened workspace
root ERROR [hosted-plugin: 12752] With stack trace: Error: Request 'setStorageValue' failed
    at Proxy.<anonymous> (http://localhost:3000/bundle.js:143539:33)
    at StorageMainImpl.push.../../packages/plugin-ext/lib/main/browser/plugin-storage.js.StorageMainImpl.$set (http://localhost:3000/56.bundle.js:9109:34)
    at RPCProtocolImpl.push.../../packages/plugin-ext/lib/common/rpc-protocol.js.RPCProtocolImpl.doInvokeHandler (http://localhost:3000/39.bundle.js:1207:23)
    at RPCProtocolImpl.push.../../packages/plugin-ext/lib/common/rpc-protocol.js.RPCProtocolImpl.invokeHandler (http://localhost:3000/39.bundle.js:1192:41)
    at RPCProtocolImpl.push.../../packages/plugin-ext/lib/common/rpc-protocol.js.RPCProtocolImpl.receiveRequest (http://localhost:3000/39.bundle.js:1156:31)
    at RPCProtocolImpl.push.../../packages/plugin-ext/lib/common/rpc-protocol.js.RPCProtocolImpl.receiveOneMessage (http://localhost:3000/39.bundle.js:1125:22)
    at http://localhost:3000/39.bundle.js:1017:109
    at http://localhost:3000/39.bundle.js:1232:17
    at http://localhost:3000/bundle.js:141091:33
    at CallbackList.../../packages/core/lib/common/event.js.CallbackList.invoke (http://localhost:3000/bundle.js:141106:39)
```
The error is in file packages\plugin-ext\src\main\node\plugins-key-value-storage.ts:59 (set method)

https://user-images.githubusercontent.com/14539932/104882035-04183480-596b-11eb-9275-88f4b506734c.mp4

I created a new extension with the use of workspaceState and got the same error but did get the string:

```
context.workspaceState.update("obj", "hi");
const msg = context.workspaceState.get("obj");
vscode.window.showInformationMessage(msg);
```
TheSQLTools also use the [same code](https://github.com/mtxr/vscode-sqltools/blob/b8938cafa59dd378d22a7ef24ecb710de51b400e/packages/plugins/connection-manager/attached-files.ts)

### Steps to Reproduce:

1. Run the server locally.
2. Install the [SQLTools](https://marketplace.visualstudio.com/items?itemName=mtxr.sqltools).
3. Make sure there is no open workspace.
4. Open a random file.
5. Close that file.
6. Get the error on the server side.

### Additional Information

- Operating System: Local
- Theia Version: master

### PR DESCRIPTION
**Pull request by danarad05**
Signed-off-by: Dan Arad dan.arad@sap.com

added code improvement for case where extension sends incorrect argument for isGlobal in StorageMain.$set method.

#### What it does
Fixed #8963 

#### How to test
1. add SQLTOOLS plugin to Theia.
2. close Workspace and open a file
3. close file and no " Cannot save data: no opened workspace" error in terminal.

#### Review checklist

- [X] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by marechal-p** [comment]
Do you know what VS Code actually does when requesting to store workspace data but no workspace is opened?

Your PR makes it store data globally when using the workspace storage API and no workspace is opened, does VS Code do the same?

File: packages/plugin-ext/src/main/browser/plugin-storage.ts [hunk 1]
```diff
@@ -43,13 +43,13 @@ export class StorageMainImpl implements StorageMain {
     }
 
     protected toKind(isGlobal: boolean): PluginStorageKind {
-        if (isGlobal) {
+        if (!isGlobal && this.workspaceService.workspace) {
+            return {
+                workspace: this.workspaceService.workspace.resource.toString(),
+                roots: this.workspaceService.tryGetRoots().map(root => root.resource.toString())
+            };
+        } else {
```
Comment (kittaakos) [inline packages/plugin-ext/src/main/browser/plugin-storage.ts:51]: The `else` is superfluous here. Just return with `undefined` after the `if` block.

Comment (danarad05) [reply packages/plugin-ext/src/main/browser/plugin-storage.ts]: You are correct. Fixed

**Comment by danarad05** [comment]
Thanks for reviewing @marechal-p. In regard to your comments:

> Do you know what VS Code actually does when requesting to store workspace data but no workspace is opened?

As far as I can tell  (after debugging VS Code also), workspaceStorage is initialized on startup in VS Code - regardless if a workspace is opened or not. 
Both workspaceStorage and globalStorage's folder path derive from User folder:

https://github.com/microsoft/vscode/blob/a699ffaee62010c4634d301da2bbdb7646b8d1da/src/vs/workbench/services/environment/browser/environmentService.ts#L124-L141

In Theia, workspaceStorage is dependent on existence of a workspace.

> Your PR makes it store data globally when using the workspace storage API and no workspace is opened, does VS Code too?

Logic for storing globally if workspace doesn't exist, was not implemented by me in this PR: in case it is global storage (isGlobal = true) then it retrieves `this.deferredGlobalDataPath`. You could argue correctly that now, instead of receiving exception, globalstorage is used - but isn't that what would happen if SQLTools would have sent correct indicator (isGlobal = true - instead of false currently) and so isn't that the wanted behavior in regard to existing workspaceStorage logic?:
https://github.com/eclipse-theia/theia/blob/6798167b50e5df4682ec218294a4eb1f3817904f/packages/plugin-ext/src/main/node/plugins-key-value-storage.ts#L91-L97

**Comment by danarad05** [comment]
@marechal-p Awaiting your continued review - if possible we would like to enter it to this week' release.
Thanks

**Comment by amiramw** [comment]
@kittaakos @marechal-p do you think we can merge this?

**Comment by kittaakos** [comment]
> @kittaakos @marechal-p do you think we can merge this?

@amiramw, I am not familiar with this logic to make a decision without further investigation. However, writing into the global storage when no workspaces are opened does not feel good. I debugged into VS Code yesterday, and I can see, the workspaceStorage is indeed initialized although no workspace was opened.

> workspaceStorage is initialized on startup in VS Code - regardless if a workspace is opened or not.

> In Theia, workspaceStorage is dependent on existence of a workspace.

It seems like this is what we need to do to be VS Code compatible. Cannot you patch/customize the`StorageMainImpl` on your side for the time being?

**Comment by danarad05** [comment]
@kittaakos 
cc: @amiramw @vince-fugnitto 

> It seems like this is what we need to do to be VS Code compatible. Cannot you patch/customize the `StorageMainImpl` on your side for the time being?

_First_, thanks @kittaakos for reviewing/replying. 

_Second_, in regard to your comment above:
I understand and agree fully that we should direct efforts to making Theia as compatible as possible to VS Code but I'm wondering if that should be done ad-hoc when a bug is found - as this PR is addressing. IMO, it's not supposed to be that while fixing a bug we should implement a change in way Theia is implemented.  IMO, ideally a Theia architect should be responsible for laying out the future vision for Theia functionality - the requirements and also not less importantly - their prioritization. 

So in short I'm asking why not fix this **bug** now so that relevant extensions could work properly and add a task to a Theia architect to contemplate future requirements? 

I would appreciate also your opinion @vince-fugnitto.

Thanks.

**Comment by kittaakos** [comment]
> why not fix this **bug** now so that relevant extensions could work properly

With the proposed change, we would write to the global storage when there is no workspace open, and that's the problem, isn't it? So you fix a bug and introduce another.

**Comment by danarad05** [comment]
> > why not fix this **bug** now so that relevant extensions could work properly
> 
> With the proposed change, we would write to the global storage when there is no workspace open, and that's the problem, isn't it? So you fix a bug and introduce another.

I don't agree - in current Theia implementation, that is the current logic - if there is a workspace/Storage - write to it, if not use global - that is underlying logic in `toKind` method. If you compare to VS Code implementation then you might be able to call it "introducing a new bug", but IMO it's not relevant as Theia wasn't and isn't totally implemented in correlation with VS Code. Currently there is an exception - IMO - for the short term solution - the exception should be fixed, for long term solution a requirement should be defined.

@vince-fugnitto ? @amiramw ?

**Comment by vince-fugnitto** [comment]
> Currently there is an exception - IMO - for the short term solution - the exception should be fixed, for long term solution a requirement should be defined.
> 
> @vince-fugnitto ? @amiramw ?

I believe I expressed in #8963 the validity of the issue, the expectation that storage be available without a currently opened workspace is incorrect (from what I understand), and while there may be an exception there is no loss of functionality. I'm not sure we should patch our side for an extension which may be doing something incorrect, like the assumption it can set storage at any state.

**Comment by danarad05** [comment]
> > Currently there is an exception - IMO - for the short term solution - the exception should be fixed, for long term solution a requirement should be defined.
> > @vince-fugnitto ? @amiramw ?
> 
> I believe I expressed in #8963 the validity of the issue, the expectation that storage be available without a currently opened workspace is incorrect (from what I understand), and while there may be an exception there is no loss of functionality. I'm not sure we should patch our side for an extension which may be doing something incorrect, like the assumption it can set storage at any state.

Fine. If there is no real issue here or no interest in fixing this issue then I think we should just close it and not waste more time on it.

**Comment by danarad05** [comment]
closing in favor of #9137.
